File: solutions/problem_0216_combination_sum_iii/solution.py
"""
Problem 216: Combination Sum III
===============================
Difficulty: Medium
Tags: Array, Backtracking
Pattern: Backtracking

Time Complexity:  O(k^n) or O(n!)
Space Complexity: O(n)
"""

class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        # Backtracking - O(2^n) or O(n!) time
        result = []
        
        def backtrack(path, start):
            result.append(path[:])
            for i in range(start, len(k)):
                path.append(k[i])
                backtrack(path, i + 1)
                path.pop()
        
        backtrack([], 0)
        return result

File: solutions/problem_0216_combination_sum_iii/solution.cpp
/*
 * Problem 216: Combination Sum III
 * ===============================
 * Difficulty: Medium
 * Tags: Array, Backtracking
 * Pattern: Backtracking
 *
 * Time Complexity:  O(k^n) or O(n!)
 * Space Complexity: O(n)
 */

#include <functional>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        // Backtracking - O(2^n) or O(n!) time
        vector<vector<int>> result;
        vector<int> path;
        function<void(int)> backtrack = [&](int start) {
            result.push_back(path);
            for (int i = start; i < (int)k.size(); i++) {
                path.push_back(k[i]);
                backtrack(i + 1);
                path.pop_back();
            }
        };
        backtrack(0);
        return result;
    }
};
